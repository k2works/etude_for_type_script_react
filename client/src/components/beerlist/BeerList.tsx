import * as React from "react";

interface IBeerListProps {
    items: any[];
}

export class BeerList extends React.Component<IBeerListProps, object> {
    public render() {
        return <ul>
            {this.props.items.map((item, index) => (
                <li key={index}>{item}</li>
            ))}
        </ul>
    };
}
