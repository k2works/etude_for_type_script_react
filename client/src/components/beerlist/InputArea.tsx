import * as React from "react";

interface IInputAreaProps {
    text: string;
    onSubmit: (name: any) => void;
    onChange: (name: any) => void;
}

export class InputArea extends React.Component<IInputAreaProps, any> {
    constructor(props: IInputAreaProps) {
        super(props);
        this.setText = this.setText.bind(this);
        this.handleClick = this.handleClick.bind(this);
    }

    public render() {
        return (
            <div>
                <input value={this.props.text} onChange={this.setText}/>
                <button onClick={this.handleClick}>Add</button>
            </div>
        );
    }

    private handleClick() {
        this.props.onSubmit(this.props.text);
    }

    private setText(event: any) {
        this.props.onChange(event.target.value);
    }
}
