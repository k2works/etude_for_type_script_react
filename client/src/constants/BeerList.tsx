export enum ActionTypes {
    ADD_ITEM = 'ADD_ITEM',
    CHANGE_ITEM = 'CHANGE_ITEM'
}
