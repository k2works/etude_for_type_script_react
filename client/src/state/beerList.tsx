export interface IState {
    beers: string[];
    text: string;
}
