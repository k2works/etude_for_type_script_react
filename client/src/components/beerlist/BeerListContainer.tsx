import * as React from "react";
import {BeerList} from "./BeerList";
import {InputArea} from "./InputArea";

interface IBeerListContainerState {
    beers: any;
}

export class BeerListContainer extends React.Component<{}, IBeerListContainerState> {
    constructor(props: any) {
        super(props);
        this.state = {
            beers: []
        };
        this.addItem = this.addItem.bind(this);
    }

    public addItem(name: any) {
        this.setState({
            beers: [].concat(this.state.beers).concat(name)
        });
    }

    public render() {
        return <div>
            <InputArea onSubmit={this.addItem}/>
            <BeerList items={this.state.beers}/>
        </div>
    }
}
