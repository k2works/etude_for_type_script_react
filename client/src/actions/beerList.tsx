import { Action } from "redux";
import { ActionTypes } from "../constants/BeerList";

interface IAddItemAction extends Action {
    type: ActionTypes.ADD_ITEM;
    payload: {
        beer: string;
    }
}

interface IChangeItemAction extends Action {
    type: ActionTypes.CHANGE_ITEM;
    payload: {
        name: string;
    }
}

export const addItem: any = (name: string): IAddItemAction => {
    return {
        type: ActionTypes.ADD_ITEM,
        payload: {
            beer: name
        }
    }
};

export const changeItem: any = (name: string): IChangeItemAction => {
    return {
        type: ActionTypes.CHANGE_ITEM,
        payload: {
            name
        }
    }
};

export type BeerListActions = IAddItemAction | IChangeItemAction;
