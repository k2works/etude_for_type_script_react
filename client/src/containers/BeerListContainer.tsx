import { connect, Dispatch } from "react-redux";
import { withRouter } from "react-router";
import * as actions from "../actions/beerList";
import { BeerListContainer } from "../components/beerlist/BeerListContainer";

export function mapStateToProps({beerList, text}: any) {
    return {
        beerList,
        text
    };
}

export function mapDispatchToProps(dispatch: Dispatch<actions.BeerListActions>) {
    return {
        addItem: (name: string) => dispatch(actions.addItem(name)),
        changeItem: (name: string) => dispatch(actions.changeItem(name))
    };
}

export default withRouter<any>(connect<any>(mapStateToProps,mapDispatchToProps)(BeerListContainer));

