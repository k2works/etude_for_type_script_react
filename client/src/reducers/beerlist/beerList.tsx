import { BeerListActions } from "../../actions/beerList";
import { ActionTypes } from "../../constants/BeerList";
import { IState } from "../../state/beerList";

const initialState = {
    beers: [],
    text: ''
};

const beerList = (state: IState = initialState, action: BeerListActions): any => {
    switch (action.type) {
        case ActionTypes.ADD_ITEM:
            return {
                ...state,
                beers: state.beers.concat(action.payload.beer),
                text: action.payload.beer
            };
        case ActionTypes.CHANGE_ITEM:
            return {
                ...state,
                beers: state.beers,
                text: state.text = action.payload.name
            };
        default:
            return state;
    }
};

export default beerList;
