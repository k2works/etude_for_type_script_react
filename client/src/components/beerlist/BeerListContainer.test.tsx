import {mount, shallow} from "enzyme";
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { IState } from "../../state/beerList";
import {BeerList} from "./BeerList";
import {BeerListContainer} from "./BeerListContainer";
import {InputArea} from "./InputArea";

describe('BeerListContainer', () => {
    xit('should render InputArea and BeerList', () => {
        const beerList: IState = {beers:[''],text:''};
        const addItemSpy = jest.fn();
        const changeItemSpy = jest.fn();
        const wrapper = shallow(<BeerListContainer addItem={addItemSpy} changeItem={changeItemSpy}  beerList={beerList}/>);
        expect(wrapper.containsAllMatchingElements([
            <InputArea key={0} onSubmit={addItemSpy} onChange={changeItemSpy} text={'test'}/>,
            <BeerList  key={1} items={[]}/>
        ])).toEqual(true);
    });

    xit('should start with an empty list', () => {
        const beerList: IState = {beers:[''],text:''};
        const addItemSpy = jest.fn();
        const changeItemSpy = jest.fn();
        const wrapper = shallow(<BeerListContainer addItem={addItemSpy} changeItem={changeItemSpy}  beerList={beerList} />);
        expect(wrapper.prop('beerList')).toEqual([]);
    });

    xit('adds items to the list', () => {
        const beerList: IState = {beers:[''],text:''};
        const addItemSpy = jest.fn();
        const changeItemSpy = jest.fn();
        const wrapper = shallow(<BeerListContainer addItem={addItemSpy} changeItem={changeItemSpy}  beerList={beerList} />);
        const instance = wrapper.instance();

        instance.props.addItem('Sam Adams');
        expect(wrapper.prop('beers')).toEqual(['Sam Adams']);
    });

    xit('passes addItem to InputArea', () => {
        const beerList: IState = {beers:[''],text:''};
        const addItemSpy = jest.fn();
        const changeItemSpy = jest.fn();
        const wrapper = shallow(<BeerListContainer addItem={addItemSpy} changeItem={changeItemSpy}  beerList={beerList} />);
        const inputArea = wrapper.find(InputArea);
        const instance = wrapper.instance();
        const addItem = instance.props.addItem;
        expect(inputArea.prop('onSubmit')).toEqual(addItem);
    });

    xit('passes a bound addItem function to InputArea', () => {
        const beerList: IState = {beers:[''],text:''};
        const addItemSpy = jest.fn();
        const changeItemSpy = jest.fn();
        const wrapper = shallow(<BeerListContainer addItem={addItemSpy} changeItem={changeItemSpy}  beerList={beerList} />);
        const inputArea = wrapper.find(InputArea);
        inputArea.prop('onSubmit')('Sam Adams');
        expect(wrapper.prop('beers')).toEqual(['Sam Adams']);
    });

    xit('renders the items', () => {
        const beerList: IState = {beers:[''],text:''};
        const addItemSpy = jest.fn();
        const changeItemSpy = jest.fn();
        const wrapper = shallow(<BeerListContainer addItem={addItemSpy} changeItem={changeItemSpy}  beerList={beerList} />);
        const addButton = wrapper.find('button');
        const input = wrapper.find('input');

        input.simulate('change', {target: { value: 'Sam Adams' }});
        addButton.simulate('click');
        input.simulate('change', {target: { value: 'Resin' }});
        addButton.simulate('click');
        expect(wrapper.find('li').length).toEqual(2);
    });

    xit('renders correctly', () => {
        const beerList: IState = {beers:[''],text:''};
        const addItemSpy = jest.fn();
        const changeItemSpy = jest.fn();
        const tree = renderer
            .create(<BeerListContainer addItem={addItemSpy} changeItem={changeItemSpy}  beerList={beerList} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('InputArea', () => {
    it('should contain an input and a button', () => {
        const addItemSpy = jest.fn();
        const changeItemSpy = jest.fn();
        const wrapper = shallow(<InputArea onSubmit={addItemSpy} onChange={changeItemSpy} text={'test'}/>);
        expect(wrapper.containsAllMatchingElements(
            [
                <input key={1} />,
                <button key={1}>Add</button>
            ]
        )).toEqual(true);
    });

    it('should accept input', () => {
        const addItemSpy = jest.fn();
        const changeItemSpy = jest.fn();
        const wrapper = mount(<InputArea onSubmit={addItemSpy} onChange={changeItemSpy} text={'Resin'}/>);
        let input = wrapper.find('input');
        input.simulate('change', {target: { value: 'Resin' }});
        expect(wrapper.prop('text')).toEqual('Resin');

        input = wrapper.find('input');
        expect(input.prop('value')).toEqual('Resin');
    });

    it('should call onSubmit when Add is clicked', () => {
        const addItemSpy = jest.fn();
        const changeItemSpy = jest.fn();
        const wrapper = shallow(<InputArea onSubmit={addItemSpy} onChange={changeItemSpy} text={'test'}/>);
        wrapper.setProps({text:'Octoberfest'});
        const addButton = wrapper.find('button');

        addButton.simulate('click');

        expect(addItemSpy).toBeCalled();
        expect(addItemSpy).toBeCalledWith('Octoberfest');
    });

    it('renders correctly', () => {
        const addItemSpy = jest.fn();
        const changeItemSpy = jest.fn();
        const tree = renderer
            .create(<InputArea  onSubmit={addItemSpy} onChange={changeItemSpy} text={''}/>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('BeerList', () => {
    it('should render zero items', () => {
        const wrapper = shallow(<BeerList items={[]}/>);
        expect(wrapper.find('li')).toHaveLength(0)
    });

    it('should render some items', () => {
        const items = ['Sam Adams', 'Resin', 'Octoberfest'];
        const wrapper = shallow(<BeerList items={items}/>);
        expect(wrapper.find('li')).toHaveLength(3)
    });

    it('renders correctly', () => {
        const tree = renderer
            .create(<BeerList  items={[]}/>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
