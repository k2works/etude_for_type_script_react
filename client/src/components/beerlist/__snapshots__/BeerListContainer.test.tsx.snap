// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeerList renders correctly 1`] = `<ul />`;

exports[`BeerListContainer renders correctly 1`] = `
<div>
  <div>
    <input
      onChange={[Function]}
      value=""
    />
    <button
      onClick={[Function]}
    >
      Add
    </button>
  </div>
  <ul />
</div>
`;

exports[`InputArea renders correctly 1`] = `
<div>
  <input
    onChange={[Function]}
    value=""
  />
  <button
    onClick={[Function]}
  >
    Add
  </button>
</div>
`;
